## Описание проекта

Компания купила крупный сервис для чтения книг по подписке.

Задачи:
 - Проанализировать базу данных.


В распоряжении 5 таблиц:

- books
- authors
- publishers
- ratings
- reviews

## Подготовка и ознакомление с таблицами

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

### Таблица books

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [3]:
query = '''
SELECT 
    *
FROM books
'''

In [4]:
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

В базе содержится информация о 1000 книг.

### Таблица authors

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

In [5]:
query = '''
SELECT 
    author_id,
    author
FROM authors

'''

In [6]:
pd.io.sql.read_sql(query, con = engine) 

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

### Таблица publishers

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [7]:
query = '''
SELECT 
    *
FROM publishers

'''

In [8]:
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

### Таблица ratings

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [9]:
query = '''
SELECT 
    *
FROM ratings

'''

In [10]:
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

### Таблица reviews

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

In [11]:
query = '''
SELECT 
    *
FROM reviews
'''

In [12]:
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## Задания 

In [13]:
# Напишем функцию для выполнения запросов.
def display(query):
    return pd.io.sql.read_sql(query, con = engine)

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [14]:
cnt_books = '''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
'''

display(cnt_books)

cnt
0  819

#### Вывод

В базе данных содержится информация о 819 книгах, дата публикации которых после 1 января 2000 года. 

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [15]:
rating = '''                              
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_qty AS reviews_qty,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(username) AS reviews_qty
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title,
    subquery_reviews.reviews_qty
ORDER BY
    avg_rating DESC
'''

display(rating)

book_name  reviews_qty  \
0                                  Homage to Catalonia          2.0   
1                                                March          2.0   
2    In the Hand of the Goddess (Song of the Liones...          2.0   
3                 Welcome to Temptation (Dempseys  #1)          2.0   
4    Stone of Farewell (Memory  Sorrow  and Thorn  #2)          1.0   
..                                                 ...          ...   
995  The World Is Flat: A Brief History of the Twen...          3.0   
996                                              Junky          2.0   
997                  His Excellency: George Washington          2.0   
998                                      Drowning Ruth          3.0   
999                               Harvesting the Heart          2.0   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

In [16]:
reviews = '''                              
SELECT
    books.title AS book_name,
    subquery_reviews.reviews_qty AS reviews_qty,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN (
        SELECT
            book_id,
            COUNT(username) AS reviews_qty
        FROM
            reviews
        GROUP BY
            book_id
        ) AS subquery_reviews ON subquery_reviews.book_id = books.book_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.title,
    subquery_reviews.reviews_qty
ORDER BY
    reviews_qty DESC
    LIMIT 10
'''

display(reviews)

book_name  reviews_qty  avg_rating
0  Disney's Beauty and the Beast (A Little Golden...          NaN    4.000000
1                          Essential Tales and Poems          NaN    4.000000
2   The Cat in the Hat and Other Dr. Seuss Favorites          NaN    5.000000
3                            The Natural Way to Draw          NaN    3.000000
4    Anne Rice's The Vampire Lestat: A Graphic Novel          NaN    3.666667
5                               Leonardo's Notebooks          NaN    4.000000
6                            Twilight (Twilight  #1)          7.0    3.662500
7                          The Giver (The Giver  #1)          6.0    3.750000
8                                           The Road          6.0    3.772727
9                                Water for Elephants          6.0    3.977273

#### Вывод

Вывели таблицу с количество обзоров и средней оценкой, с сортировкой по средней оценке.
Вторая таблица показывает, что у нас по 6 книгам отсутствуют обзоры. И максимальное количество обзоров на книги 7.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [17]:
publisher_book = '''
SELECT
    publishers.publisher,
    COUNT(book_id) AS books_cnt
FROM
    books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE 
    num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_cnt DESC
LIMIT 1
'''
display(publisher_book)

publisher  books_cnt
0  Penguin Books         42

#### Вывод

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги. 

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [18]:
best_avtor = '''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
    LEFT JOIN (
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_cnt >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
'''
display(best_avtor)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

#### Вывод

Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré. Средняя оценка ее книг составляет 4.28 балла.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
avg_reviews = '''
WITH
    ratings_subquery AS (
        SELECT
            ratings.username,
            COUNT(rating_id) AS ratings_cnt
        FROM
            ratings
        GROUP BY
            ratings.username
        HAVING
            COUNT(rating_id) > 50
        ),
    reviews_subquery AS (
        SELECT
            reviews.username,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            reviews.username
    )
SELECT
    AVG(reviews_cnt) as avg_reviews_cnt
FROM
    ratings_subquery
    LEFT JOIN reviews_subquery ON reviews_subquery.username = ratings_subquery.username
'''
display(avg_reviews)

avg_reviews_cnt
0        24.333333

#### Вывод

В БД присутствует информация об активных пользователях, оставивших более 50 оценок. В среднем такие пользователи оставляют 24.3 обзоров на книгу.

## Общий вывод

В ходе данного проекта была проанализирована база данных крупного сервиса для чтения книг по подписке. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. В резулате анализа получены следующие результаты:

- В БД нового сервиса содержится информация о 1000 книг.
- В БД содержится информация о 636 авторах, 340 издательствах, 6456 оценках книг и о 2793 обзорах на книги.
- В базе данных содержится информация о 819 книгах, дата публикации которых после 1 января 2000 года.
- Вывели таблицу с количество обзоров и средней оценкой, с сортировкой по средней оценке.
- Вторая таблица показывает, что у нас по 6 книгам отсутствуют обзоры.
- Максимальное количество обзоров на книги 7.
- Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги.
- Самая высокая средняя оценка книг у автора J.K. Rowling/Mary GrandPré. Средняя оценка ее книг составляет 4.28 балла.
- В БД присутствует информация об активных пользователях, оставивших более 50 оценок. В среднем такие пользователи оставляют       24.3 обзоров на книгу.
